### DISCUSSION TITLE
Nullifying stdout for split communicators

### DISCUSSION THREAD

**Discussion by RonRahaman** [idaholab/moose#16240] [Q&A General]
Hi all, let me know if I should move this thread to a libmesh forum.  It's relevant to our external apps in MOOSE, so I thought I'd mention it here.  

Our MOOSE app consists of some ExternalProblems from other single-physics apps.  Currently, each single-physics app is running on the global comm.  In the near future, we'd like to run each single-physics app on a comm that was split from the global comm.  

In the split comm situation, we'd like a better way to handle stdout.  The problem is this.  Suppose we're running MOOSE such that stdout is nullified for every rank except global rank 0, as done here:
 https://github.com/libMesh/libmesh/blob/a34662c8df2811388c6760bf537829ea62b181b9/src/base/libmesh.C#L622-L627

Now suppose a single-physics app is run on a split comm that doesn't include global rank 0.  In that case, that app would not produce any stdout.  

We're aware of the nuclear `--keep-cout` option, but to keep stdout from getting out of control, we'd like some more fine-grained control.  Maybe if we could specify particular ranks to keep stdout?  We'd like to do that in the code, rather than through a config file, since the code would need to discover the roots of the split comms and then keep stdout for those.  Definitely open to other options!

**Comment by aeslaughter** [comment]
@friedmud Do you know the answer to this off the top of your head? Sorry, we missed this question. We are still getting used to this tool.

**Comment by GiudGiud** [comment]
@friedmud since you said you would respond to notifications from now on ^^
I think we could support --keep-cout #rank_we_care_about
Is there still interest for this?

**Reply by loganharbour** [reply]
I like this. It should support a vector of ranks. I know I’d use it.

**Reply by YaqiWang** [reply]
The syntax can be `0,1,3-10,100`.

**Reply by GiudGiud** [reply]
anyone want to work on this?

**Reply by loganharbour** [reply]
I can give it a crack real quick.

**Reply by loganharbour** [reply]
Actually - `--keep-cout` is managed by libMesh. Ping @roystgnr for his thoughts

**Reply by roystgnr** [reply]
This sounds good, as long as we only interpret numeric arguments (including commas and dashes) and we're careful to parse and ignore anything alphabetical?  We don't want `file_argument.x --keep-cout input=myfile.e` to suddenly get confused and die because we thought `input=myfile.e` was a vector of ranks but it failed to parse as one, we want to just maintain the old "all ranks" behavior in that case.

I do worry a little about `numeric_argument.x --keep_cout 64` ... but I don't think I've ever seen any applications behave that way, with a numeric argument that wasn't either identified via a fixed argument-vector position (in which case nobody would put `--` arguments in front of it rather than behind it) or attached to a `--identifier` or `identifier=` argument, so we can probably ignore that worry.

**Reply by friedmud** [reply]
I like this option... but we also need an API to set this programmatically.
